### ISSUE TITLE
About the attention model

### ISSUE THREAD

**Issue by skyw** [google/seq2seq#127]
From the code and the graph, it looks attention is only applied to the last decoder layer?
Am I reading it wrong?

**Comment by dennybritz** [comment]
Yes, that's correct.

**Comment by okuchaiev** [comment]
@dennybritz Could you please comment on different ways of consuming attention context?
As @skyw mentioned in current code it is fed to the last decoder layer, and it seems like to the FC layer that goes into softmax as well. 
GNMT paper seems to use attention at each layer of the decoder.
And original TF seq2seq legacy model seems to feed it only to the bottom layer of decoder: [https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/legacy_seq2seq/python/ops/seq2seq.py#L683](TF seq2seq legacyl)

**Comment by dennybritz** [comment]
There are lots of different ways to do attention. The standard way (what this code does and also what the legacy seq2seq does is as follows):

- Calculate attention based on the last layer of the decoder, add/concat it to the decoder state and feed that into the softmax.
- Append the calculated attention context to the *next* input so that the model can learn what it has previously attended to.

> And original TF seq2seq legacy model seems to feed it only to the bottom layer of decoder: 

It also feeds it into the softmax. I think it does the same as this code and the same as describe above.

GNMT does a bunch of special things, but a lot of it is for performance reasons. I think it's unclear how much it actually helps in making better predictions.

**Comment by skyw** [comment]
I believe both GNMT and legacy seq2seq feed attention to all decoder layers. The difference is GNMT only sends the bottom layer's state as query, while legacy seq2seq sends each layer's own state as query.

**Comment by dennybritz** [comment]
> I believe both GNMT and legacy seq2seq feed attention to all decoder layers

I could be wrong but I don't think so. All standard models only put attention on the last layer. For example, in the legacy code attention is added [in this line](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/legacy_seq2seq/python/ops/seq2seq.py#L695), which is the output of the (possible stacked) RNN cell. Thus it is added at the top.
